descriptor/qsarmolecular/src/main/java/org/openscience/cdk/qsar/descriptors/molecular/FractionalCSP3Descriptor.java

```diff
@@ -0,0 +1,137 @@
+ */
+
+package org.openscience.cdk.qsar.descriptors.molecular;
+
+import org.openscience.cdk.exception.CDKException;
+import org.openscience.cdk.interfaces.IAtom;
+import org.openscience.cdk.interfaces.IAtomContainer;
+import org.openscience.cdk.interfaces.IAtomType.Hybridization;
+import org.openscience.cdk.qsar.AbstractMolecularDescriptor;
+import org.openscience.cdk.qsar.DescriptorSpecification;
+import org.openscience.cdk.qsar.DescriptorValue;
+import org.openscience.cdk.qsar.IMolecularDescriptor;
+import org.openscience.cdk.qsar.result.DoubleResult;
+import org.openscience.cdk.qsar.result.DoubleResultType;
+import org.openscience.cdk.qsar.result.IDescriptorResult;
+import org.openscience.cdk.tools.manipulator.AtomContainerManipulator;
+
+/**
+ * An implementation of the Fractional CSP3 descriptor characterizing non-flatness of a molecule.
+ * 
+ * This descriptor returns a single double value, labeled as "Fsp3"
+ * 
+ * @author Kazuya Ujihara
+ * @cdk.module qsarmolecular
+ * @cdk.dictref qsar-descriptors:Fsp3
+ * @cdk.githash
+ */
+public class FractionalCSP3Descriptor extends AbstractMolecularDescriptor implements IMolecularDescriptor {
+    public FractionalCSP3Descriptor() { }
+    
+
+    /** {@inheritDoc} */
+    @Override
+    public DescriptorSpecification getSpecification() {
+        return new DescriptorSpecification("http://www.blueobelisk.org/ontologies/chemoinformatics-algorithms/#Fsp3",
+                this.getClass().getName(), "The Chemistry Development Kit");
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String[] getParameterNames() {
+        return new String[0];
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Object getParameterType(String name) {
+        return null;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void setParameters(Object[] params) throws CDKException {
+        if (params.length != 0) {
+            throw new CDKException("The FractionalCSP3Descriptor expects zero parameters");
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Object[] getParameters() {
+        return new Object[0];
+    }
+
+    /**
+     * Calculates the Fsp<sup>3</sup> descriptor value for the given {@link IAtomContainer}.
+     *
+     * @param container An {@link org.openscience.cdk.interfaces.IAtomContainer} for which this descriptor
+     *                  should be calculated
+     * @return An object of {@link org.openscience.cdk.qsar.DescriptorValue} that contains the
+     *         calculated Fsp<sup>3</sup> descriptor value
+     */
+    @Override
+    public DescriptorValue calculate(IAtomContainer mol) {
+        try {
+            mol = mol.clone();
+        } catch (CloneNotSupportedException ex) {
+        }
+        
+        int nC = 0;
+        int nCSP3 = 0;
+        DoubleResult result;
+        try {
+            AtomContainerManipulator.percieveAtomTypesAndConfigureAtoms(mol);
```

Comment:
I would suggest to just perceive the atom types, but not apply them to the molecule... then you do not have to `clone()` the IAtomContainer, making the whole calculation a lot faster.

Reply:
(Look for the atom type perception method that returns the list of perceived atom types as a list...)

Reply:
I put it in my comment :-), but you can do even better by avoiding that all together.

Reply:
Yes, that may be tempting indeed. I slightly prefer to keep things modular and not mix perception with counting, but that's just personal taste...

Reply:
I updated to avoid using clone. 
Thanks for your suggestions.

Reply:
@johnmay, yes, different algorithms need different atom type sets. The point is, we should not have multiple perception codes to perceive the *same* set of atom types. That was the situation we had in the early days of the CDK.